src/java.base/share/classes/java/lang/String.java

```diff
@@ -2177,10 +2177,14 @@ public boolean regionMatches(int toffset, String other, int ooffset, int len) {
         if (len <= 0) {
            return true;
         }
+        byte coder = coder();
+        byte otherCoder = other.coder();
         byte[] tv = value;
         byte[] ov = other.value;
-        byte coder = coder();
-        if (coder == other.coder()) {
+        if (coder == otherCoder) {
+            if ((ooffset | toffset) == 0 && len == (tv.length >> coder) && ov.length == tv.length) {
```

Comment:
It can be, depending the probability of happening for both and if/where they are inlined: conditions easy to predict are very cheap, but if not, can have some performance penalties. 
An `|` between register is super cheap, instead: the same tecnique has been used at https://github.com/openjdk/jdk/blob/a26f7c03c72e4efe6d3219d294294c554aebc631/src/hotspot/share/utilities/copy.cpp#L215 which is supposed to be another hot-path, probably for similar reasons.